### DISCUSSION TITLE
Prototyping new XState based UI architecture

### DISCUSSION THREAD

**Discussion by UberMouse** [statelyai/xstate#1436] [General] [answered]
I've been spending the last few days looking at how to make a hierarchical, multi-machine, React backed view system for XState. I've now got a working prototype powering a ToDo MVC app so I'm interested in hearing what other people think about the pattern.

Fundamentally it's about automatically mapping a 1:1 representation of your machine hierarchy onto a hierarchy of React views. This is accomplished by tagging each machine with some meta attributes, `view` and optionally `selectors` and `actions`. A view for a machine is passed as props `slots`, which are any invoked/spawned child machines matching an id format (so you can have machines that have no UI representation), `selectors` and `actions`. `selectors` and `actions` are the result of calling the `meta.selectors`, `meta.actions` functions with the machines context/state.matches and selectors/send respectively. This helps to minimize the coupling between the view and the machine. `slots` are the child machines wrapped into React components to allow you to render them wherever you want in the machines view, "slotting" them in. This allows the machine to not have to be concerned with how it's child machines are rendering, just _where_ it wants them to be rendered. There is also functionality to broadcast events to all machines, instead of sending them to a specific machine.

The "toplevel" machine which is responsible for invoking any child machines (which in turn can invoke further child machines for managing subsections of their UI, creating the hierarchy) has a React component constructed from it which kicks off the whole process, mounts the machines, creates the views, renders them recursively etc when it is mounted into the DOM.

All of this is a bit of hacky, likely unperformant, mess in `useViewService.js`. `AppMachine` invokes `TodosMachine` into it's `Screen` slot (by invoking the machine with the id `screen-slot`) which in turn spawns a `TodoMachine` actor for each created Todo into the `Todo` slot (which since it's a slot representing multiple spawned actors is wrapped into a container component which renders the view for each actor internally).

tl;dr

- Machines get tagged with meta attributes to describe how to render them
- View gets machine state/interaction exposed via selectors and actions props. Selector function gets passed machine `context` and `state.matches`, actions function gets passed `send` and the built `selectors`. Child views are exposed via the `slots` prop to allow building up the view tree.
- Machines invoke/spawn child machines with a specific id format ([slotName]-slot or id-[slotName]-slots depending on whether it's a single machine or a collection of machines)
- A React component is constructed based off the top-level machine which kicks off the process and recursively renders the views for its slotted machines
- Communication is either done via child->parent or by broadcasting an event to all machines. Ideally I would also like a way to send events to a specific machine by id or set of machines (like all the TodoMachine actors) but not sure how to do that atm since the ids are the slots instead of any machine id. But maybe sending them to the slots makes sense 🤔 

And here's a codesandbox demonstrating it https://codesandbox.io/s/friendly-black-fym5j?file=/src/TodosMachine.js:4031-4039

This was inspired by a [Masters Thesis](https://links-lang.org/papers/mscs/Master_Thesis_Thomas_Weber_1450761.pdf) by Thomas Weber on representing a UI with actors.

**Comment by UberMouse** [comment]
I also have plenty of unsolved problems with how to make this performant. I'm suspecting I may need a totally different approach. There is quite a bit going on and more hooks run than I think are required. Getting all of the views to re-run when their data might have changed proved to be tricky, ensuring the constructed view components didn't get rebuilt is a bit of a hack as well.

I've especially had difficulty determining if the children of a service have changed as it's a mutable map with no way to hook into an event to notify about it that I can see. I've had to resort to adding an onTransition handler to each child of a service and force rerendering the component if any of them change. As well as listening for the `done.invoke` event to remove that child from my children references. Plenty of improvements could be made there I'm sure.

**Comment by tivac** [comment]
I wish I understood this better, but it's a really interesting approach. All the react-isms make it tricky to follow for my svelte-addled 🧠 but the idea of making every child component its own invoked machine has a certain purity to it that I really like. We've fiddled with creating small statecharts for components where it made sense but so far have tended to work in singular large statecharts vs many small ones. I'd like to balance that out a bit more over time but paying off tech debt is always a tricky thing to prioritize.

I definitely ran into some of the same `children` tracking issues when I was putting together [`xstate-component-tree`](https://github.com/tivac/xstate-component-tree/), I made the diffing as efficient as I could but there's no easy win there w/o deeper xstate integration that I could come up with :(

**Reply by UberMouse** [reply]
It doesn't help the code is a mess 😂 primarily interested in discussing the concept demonstrated in the machines themselves. I think it has potential but it's hard to know what issues it will run into modelling a more complex UI with more interdependencies. Without actually just doing that, which I will likely be doing this week at work. Just need to improve the "framework" a bit first. Hoping to eventually release a package for this, ideally with a system to allow a React version and a Vue version etc.

**Comment by 1FreeInhabitant** [comment]
I love this... And am in the middle of a similar decision...

How to use xState as an auto UI builder... Theme building, etc. Following this thread... Thanks.

**Reply by UberMouse** [reply]
[Here](https://codesandbox.io/s/pensive-sammet-y1gk0?file=/src/TodosMachine.js) is my latest progress. I'm happy with the design of the machines + views + selectors + actions, but the underlying framework is a big mess (all the code in `useViewservice.js`). My next step is to take the concept and flesh it out properly in our codebase at work in prep for re-developing our existing UI on it which will further stress test the concept and show me what other work needs to be done on it to make it something truly useful. And whether it can be made performant enough, I've got some concerns with that but I think it can be solved with a better design.

My hope is to eventually open-source it as a stand-alone package.

**Comment by fraserdrops** [comment]
This reminds me of @rjdestigter 's experiment here https://spectrum.chat/statecharts/general/experimenting-with-putting-xstate-in-charge-of-react-rather-than-vice-versa~67bd71e8-7f0d-40e3-aaeb-11e5d167217a

**Comment by redbar0n** [comment]
>  Fundamentally it's about automatically mapping a 1:1 representation of your machine hierarchy onto a hierarchy of React views.

>  I'm suspecting I may need a totally different approach.

> more hooks run than I think are required

What I envision, is [a vertical slice MVC-widget based architecture](https://gist.github.com/redbar0n/545e94cc4e0ca1dfc922459251fd7c5a) where:

1. User actions in the View triggers actions on corresponding XState machine.
2. XState machine changes the route state [update: actually the app state, which would change the URL].
3. React will automatically re-render the View based on the new route state. See [routing ideas](https://gist.github.com/redbar0n/f6ec12264ff9d58e243cc516e4e3f41b) on how to achieve that.

You don't need XState to directly drive the rendering in React (with the likes of [xstate-component-tree](https://github.com/tivac/xstate-component-tree) or [xstate-first-react](https://spectrum.chat/statecharts/general/experimenting-with-putting-xstate-in-charge-of-react-rather-than-vice-versa~67bd71e8-7f0d-40e3-aaeb-11e5d167217a) or similar). It suffices that XState changes the route state, which React components pick up (through prop change or a hook) and uses it to re-renders itself.

Rendering thus becomes a side-effect (like it always should have been...).

Rendering ought to be _an end-result_, not the start of a list of potential other side-effects (I'm looking at you, `useEffect`).

There's an open question, with some sub-questions: Do you need to model the router as a global XState machine in itself, or can it be a simple route object? If you model it as an XState machine, will all the various combinations of states grow out of control? And will it be harder (than a simple route object) to pass the entire machine through the React render tree? Also, if you model the route state as a state machine router, you'd have to define the component hierarchy twice: once in the render tree, and once in the state tree (connecting all the machines). That's cumbersome, and a bit of boilerplate. Not to mention trying to keep them in sync, over time... Or trying to code-split this router.

Let me know what you think.

**Comment by UberMouse** [comment, marked as answer]
I've just open sourced the productionized version of this prototype we have been using at Koordinates for the past two years https://github.com/koordinates/xstate-tree
